src/operators/ai.onnx/ConvTranspose/11/execute_operator__ai_onnx__convtranspose__11__T_tensor_float.c

```diff
@@ -0,0 +1,193 @@
+    // size_t n_kernel_shape = op_ctx->n_kernel_shape;
+    // int64_t* kernel_shape = op_ctx->kernel_shape;
+    // size_t n_output_padding = op_ctx->n_output_padding;
+    // int64_t* output_padding = op_ctx->output_padding;
+    // size_t n_output_shape = op_ctx->n_output_shape;
+    // int64_t* output_shape = op_ctx->output_shape;
+    // size_t n_pads = op_ctx->n_pads;
+    // int64_t* pads = op_ctx->pads;
+    // size_t n_strides = op_ctx->n_strides;
+    // int64_t* strides = op_ctx->strides;
+
+    // TRACE_VAR(2, true, auto_pad, "\"%s\"");
+    // TRACE_ARRAY(2, true, dilations, , n_dilations, "%" PRId64);
+    // TRACE_VAR(2, true, group, "%" PRId64);
+    // TRACE_ARRAY(2, true, kernel_shape, , n_kernel_shape, "%" PRId64);
+    // TRACE_ARRAY(2, true, output_padding, , n_output_padding, "%" PRId64);
+    // TRACE_ARRAY(2, true, output_shape, , n_output_shape, "%" PRId64);
+    // TRACE_ARRAY(2, true, pads, , n_pads, "%" PRId64);
+    // TRACE_ARRAY(2, true, strides, , n_strides, "%" PRId64);
+
+    Onnx__TensorProto *o_Y = searchOutputByName(ctx, 0);
+    printf("%p\n", o_Y->float_data);
+
+    // TRACE_TENSOR(2, true, o_Y);
+
+    /* DO CALCULATION HERE */
+
+    const int inputSizeX = i_X->dims[3];
+    const int inputSizeY = i_X->dims[2];
+    const int inputChannels = i_X->dims[1];
+
+    const float *input = i_X->float_data;
+    const float *weights = i_W->float_data;
+
+    const int kernelSizeX = i_W->dims[3];
+    const int kernelSizeY = i_W->dims[2];
+    const int outputChannels = i_W->dims[1];
+
+    int strideX = 1;
+    int strideY = 1;
+
+    if(a_strides != NULL) {
+        strideX = a_strides->ints[1];
+        strideY = a_strides->ints[0];
+    }
+
+    int dilationsX = 1;
+    int dilationsY = 1;
+
+    if(a_dilations != NULL) {
+        dilationsX = a_dilations->ints[1];
+        dilationsY = a_dilations->ints[0];
+    }
+
+    int padStartY = 0;
+    int padStartX = 0;
+    int padEndY = 0;
+    int padEndX = 0;
+
+    if(a_pads != NULL) {
+        assert(a_pads->n_ints == 4);
+
+        padStartY = a_pads->ints[0];
+        padStartX = a_pads->ints[1];
+        padEndY = a_pads->ints[2];
+        padEndX = a_pads->ints[3];
+    }
+
+    const int outputSizeX = calcOutputSize(inputSizeX, kernelSizeX, strideX, dilationsX, 0, 0);
+    const int outputSizeY = calcOutputSize(inputSizeY, kernelSizeY, strideY, dilationsY, 0, 0);
+
+    const int outputSizePaddedX = calcOutputSize(inputSizeX, kernelSizeX, strideX, dilationsX, padStartX, padEndX);
+    const int outputSizePaddedY = calcOutputSize(inputSizeY, kernelSizeY, strideY, dilationsY, padStartY, padEndY);
+
+    //if badding is needed we need an extra tensor and applay padding later
+    float* output = NULL;
+    if(outputSizeX * outputSizeY != outputSizePaddedX * outputSizePaddedY) {
+        output = (float*)malloc(outputSizeX * outputSizeY * outputChannels * sizeof(float));
+    } else {
+        output = o_Y->float_data;
+    }
```

Comment:
but why? why not have the padding in the index calculation?

Reply:
That would involve multiple if statements to not write over the border. The problem with the padding in ConvTranspose2d is that the padding makes the result smaller (the lines get cut off). Any ideas how we could solve this better?

Reply:
> That would involve multiple if statements to not write over the border.

Do you think this will be a performance problem? I think the kernel detection in gcc should optimize this pretty well.
At least I would allocate this buffer inside the prepare function, so the executer does no malloc.

Reply:
I don't know anything about the kernel detection in gcc. The final convolution are six for loops for a 4d tensor adding additional statements have the potential for a huge impact to make it slower. Maybe we should benchmark it.

Reply:
yeah I was more concerned about copying these large amounts of data as about the padding edge case.
but whatever, if it is slow we can optimize it later on

Reply:
Isn't this operator very familiar with the conv operator? we could benchmark your implementation against our already existing one
[execute_operator__ai_onnx__conv__11__T_tensor_float](https://github.com/alrevuelta/cONNXr/blob/master/src/operators/ai.onnx/Conv/11/execute_operator__ai_onnx__conv__11__T_tensor_float.c)

Reply:
It is kinda similar because it does the exact opposite than conv